### PR TITLE
Add file cache in local database

### PR DESCRIPTION
**Pull request by kosssi**

### CODE REVIEW

File: src/app/lib/file_cache_handler.coffee [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            callback()
+
+
+    getFileName: (cozyFile) ->
+        log.debug 'getFileName'
+
+        return encodeURIComponent cozyFile.name if cozyFile.name
+
+        log.warn JSON.stringify cozyFile
+        new Error 'cozyFile hasn\'t name field'
```
Comment (nono) [inline src/app/lib/file_cache_handler.coffee]: @kosssi you don't `throw` the error?

File: src/app/lib/file_cache_handler.coffee [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        log.debug "getBinary"
+
+        folderName = @getFolderName cozyFile
+        return callback folderName unless typeof folderName is 'string'
+        fileName = @getFileName cozyFile
+        return callback fileName unless typeof fileName is 'string'
+
+        fs.getOrCreateSubFolder @downloads, folderName, (err, binaryFolder) =>
+            if err and err.code? isnt FileError.PATH_EXISTS_ERR
+                ret urn callback err
```
Comment (nono) [inline src/app/lib/file_cache_handler.coffee]: There is a space between `ret` and `urn`

File: src/app/replicator/change/change_file_handler.coffee [hunk 3]
```diff
@@ -29,18 +32,14 @@ module.exports = class ChangeFileHandler
...
+            return callback() unless @fileCacheHandler.isCached doc
 
-            # Entry is false if this file isn't cached.
-            return callback() unless entry
-
-            if entry.name isnt @_fileToEntryName doc
-                @_update doc, callback
+            unless @fileCacheHandler.isSameBinary doc
+                @_download doc, callback
+            else unless @fileCacheHandler.isSameName doc
```
Comment (nono) [inline src/app/replicator/change/change_file_handler.coffee]: unless ... else unless ... is really hard to read and understand

**Comment by nono** [comment]
The build on travis is red

File: .travis.yml [hunk 4]
```diff
@@ -4,16 +4,20 @@ language: android
...
+env:
+  - TRAVIS_NODE_VERSION="4"
+
 android:
   components:
     - android-22
     - build-tools-22.0.1
 
 install:
+  - rm -rf ~/.nvm && git clone https://github.com/creationix/nvm.git ~/.nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source ~/.nvm/nvm.sh && nvm install $TRAVIS_NODE_VERSION
```
Comment (nono) [inline .travis.yml]: Why do you remove and reinstall nvm?

File: src/app/lib/file_cache_handler.coffee [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+instance = null
+
+
+module.exports = class FileCacheHandler
+
+
+    constructor: (@localDb, @replicateDb, @requestCozy) ->
+        return instance if instance
+        instance = @
+        @cache = {}
```
Comment (nono) [inline src/app/lib/file_cache_handler.coffee]: It looks like you are using `@cache` as an hash. You should use `@cache = Object.create(null)` if it is the case. For example `@cache['toString']` will be truthy if instanciated as `{}`, but falsy (as expected) if instanciated as `Object.create(null)`.

See https://glebbahmutov.com/blog/sweet-naked-objects/ and http://www.devthought.com/2012/01/18/an-object-is-not-a-hash/

**Comment by kosssi** [comment]
done @nono

**Pull request merged**